PostHog/posthog
# feat: verify email addresses on signup

## Problem

We don't currently verify email addresses on signup, but we should. See: https://posthog.slack.com/archives/C0428JHABK2/p1670337877813209



## Changes

This _should_ only impact signup/login on cloud. If someone is self-hosting I don't think we need to worry about this. The important bits should also all be behind a feature flag, lmk if I missed anything.

* When someone signs up, they are not immediately logged in. They are sent to a `/verify_email` page telling them that they need to..... verify their email.
  - ![image](https://user-images.githubusercontent.com/18598166/211718513-6be6fec6-ec21-4817-9539-3dece7d8050c.png)
* An email is sent out with a link to verify (hopefully images won't be broken on prod and when using the correct email account?)
  - ![image](https://user-images.githubusercontent.com/18598166/211719292-d8abe7d6-3873-4bec-9d21-0734c91ac198.png)
* When they click the link, it verifies their token, logs them in, and redirects to `/`
* If they try to log in but `user.is_email_verified == False`, another verification email is sent and they get an error telling them they need to verify their email address
* If they lose the verification email they can request a new one with a lil button or by attempting to log in again
* Because user accounts are created before the email is verified, we need to differentiate people who join orgs as those who are verified or not. So, now there is a tag on unverified members in the org settings stating as much.
  - ![image](https://user-images.githubusercontent.com/18598166/217140098-b8d9a4f4-aad0-462c-960a-2b6b83d32e2b.png)
* If someone has an unverified email address, we show a notice for them saying that they need to verify. 
  - ![image](https://user-images.githubusercontent.com/18598166/217149493-75f3edd1-0607-42e9-936b-7b5e842cb176.png)
 - If an account exists with username/pw, and someone logs in with SSO, the PW is stripped from the account.
 - Invites also require email verification
 - Existing users are allowed to log in, but upon login will see a notice asking them to verify their email address. We can lock them out at a later date after they've had an opportunity to verify. 
 - If someone changes their email address they have to verify the new address before it is actually set for their account. Until they verify there is a tag on the settings page saying it is pending verification. 
  - <img width="559" alt="image" src="https://user-images.githubusercontent.com/18598166/219967410-144e52b5-c1a7-469e-bc68-ebafd947801b.png">
  - There currently isn't a way to remove the pending email because I don't think we will run into that need too frequently. If they mis-type the email, or decide they don't want to change it, they can verify one they have access to (eg me+1@mycompany.com) and then change and verify again the original one (eg me@mycompany.com).





👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Tests and manually.

**To test this yourself:**

1. Pull down this branch.
2. Message me on Slack for some env vars you can use to get email sending working locally (I have a SendGrid account I've been using for testing the email sending)
3. In `posthog/utils.py` `get_self_capture_api_token` return the API key for a local project you are using (so you can set the feature flag that turns this on). (this is hacky, is there a better way to define which project API key to use?)
4. Sync feature flags with `DEBUG=1 python3 manage.py sync_feature_flags` so that the flag will be created and turned on.
5. Set `is_cloud()` to `return True`. (also hacky, is there a better way to do this? maybe an env var?)
6. Finally, go to `http://localhost:8000/signup` and you should be able to create a new account, get the verification UI and emails (check your spam folder as they will likely land in there), and do the things!
